flink-end-to-end-tests/src/main/java/org/apache/flink/streaming/tests/StickyAllocationAndLocalRecoveryTestJob.java

```diff
@@ -0,0 +1,451 @@
+ */
+public class StickyAllocationAndLocalRecoveryTestJob {
+
+	private static final Logger LOG = LoggerFactory.getLogger(StickyAllocationAndLocalRecoveryTestJob.class);
+
+	public static void main(String[] args) throws Exception {
+
+		final ParameterTool pt = ParameterTool.fromArgs(args);
+
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		env.setParallelism(pt.getInt("parallelism", 1));
+		env.setMaxParallelism(pt.getInt("maxParallelism", pt.getInt("parallelism", 1)));
+		env.enableCheckpointing(pt.getInt("checkpointInterval", 1000));
+		env.setRestartStrategy(RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, pt.getInt("restartDelay", 0)));
+		if (pt.has("externalizedCheckpoints") && pt.getBoolean("externalizedCheckpoints", false)) {
+			env.getCheckpointConfig().enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);
+		}
+
+		String stateBackend = pt.get("stateBackend", "file");
+		String checkpointDir = pt.getRequired("checkpointDir");
+
+		boolean killJvmOnFail = pt.getBoolean("killJvmOnFail", false);
+
+		if ("file".equals(stateBackend)) {
+			boolean asyncCheckpoints = pt.getBoolean("asyncCheckpoints", false);
+			env.setStateBackend(new FsStateBackend(checkpointDir, asyncCheckpoints));
+		} else if ("rocks".equals(stateBackend)) {
+			boolean incrementalCheckpoints = pt.getBoolean("incrementalCheckpoints", false);
+			env.setStateBackend(new RocksDBStateBackend(checkpointDir, incrementalCheckpoints));
+		} else {
+			throw new IllegalArgumentException("Unknown backend: " + stateBackend);
+		}
+
+		// make parameters available in the web interface
+		env.getConfig().setGlobalJobParameters(pt);
+
+		// delay to throttle down the production of the source
+		long delay = pt.has("delay") ? pt.getLong("delay") : 0L;
+
+		// the maximum number of attempts, before the job finishes with success
+		int maxAttempts = pt.has("maxAttempts") ? pt.getInt("maxAttempts") : 3;
+
+		// size of one artificial value
+		int valueSize = pt.has("valueSize") ? pt.getInt("valueSize") : 10;
+
+		env.addSource(new RandomLongSource(maxAttempts, delay))
+			.keyBy((KeySelector<Long, Long>) aLong -> aLong)
+			.flatMap(new StateCreatingFlatMap(valueSize, killJvmOnFail))
+			.map((MapFunction<String, String>) value -> value)
+			.addSink(new PrintSinkFunction<>());
+
+		env.execute("Sticky Allocation And Local Recovery Test");
+	}
+
+	/**
+	 * Source function that produces a long sequence.
+	 */
+	private static final class RandomLongSource extends RichSourceFunction<Long> implements CheckpointedFunction {
+
+		private static final long serialVersionUID = 1L;
+
+		/**
+		 * Generator delay between two events.
+		 */
+		final long delay;
+
+		/**
+		 * Maximum restarts before shutting down this source.
+		 */
+		final int maxAttempts;
+
+		/**
+		 * State that holds the current key for recovery.
+		 */
+		ListState<Long> sourceCurrentKeyState;
```

Comment:
This field is not serializable and should thus be marked `transient`.

Reply:
👍